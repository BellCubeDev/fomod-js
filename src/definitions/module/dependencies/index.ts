export * from './Dependency';
export * from "./DependenciesGroup";
export * from "./FileDependency";
export * from "./FlagDependency";
export * from "./VersionDependency";
