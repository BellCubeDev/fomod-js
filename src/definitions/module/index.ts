export * from './dependencies';
export * from './Fomod';
export * from './Group';
export * from './Install';
export * from './Option';
export * from './Step';
