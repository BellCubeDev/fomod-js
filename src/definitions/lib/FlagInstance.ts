import { Option } from '../module/Option';

export interface FlagInstances {
    all: Set<FlagInstance<boolean, boolean>>;
    byName: Map<string|Option<boolean>, Set<FlagInstance<boolean, boolean>>>;
}

/** A [weak map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakMap) of documents to flag instances within that document.
 *
 * This map allows for quick access to all flag instances within a document as well as all flag instances with a given name.
 */
export const FlagInstancesByDocument = new WeakMap<Document, FlagInstances>();

export class FlagInstance<TIsOption extends boolean, TWrite extends (TIsOption extends true ? false : boolean)> {

    get name() { return this._name; }
    set name(value: TIsOption extends true ? Option<boolean> : string) {
        if (this.name === value) {
            this._name = value;
            return;
        }

        this.documents.forEach(document => {
            const instances = FlagInstancesByDocument.get(document);
            if (!instances) return;

            instances.byName.get(this.name)?.delete(this);

            let instancesByName = instances.byName.get(value);
            if (!instancesByName) {
                instancesByName = new Set();
                instances.byName.set(value, instancesByName);
            }

            instancesByName.add(this);
        });

        this._name = value;
    }

    /** A list of documents this flag is a part of */
    documents: Set<Document> = new Set();

    constructor(_name: TIsOption extends true ? never : string, usedValue: string, write: TWrite, document?: Document)
    constructor(_name: TIsOption extends true ? Option<boolean> : never, usedValue: true, write: TWrite, document?: Document)
    constructor(
        /** The name of the flag this instance refers to
         *
         * Automatically updates the flag instance map when changed.
        */
        private _name: TIsOption extends true ? Option<boolean> : string,


        /** The value being checked/set by this instance
         *
         * When the flag represents a traditional key-value, this will be a string. When the flag represents an option's select state, this will be `true`.
         */
        public usedValue: TIsOption extends true? true : string,

        /** Whether this instance is reading or writing the flag value
         *
         * When the flag represents an option's select state, this mist be `false` (since you cannot set the value of an option's select state)
         */
        public write: TWrite,

        document?: Document
    ) {
        if (typeof _name === 'string' && typeof usedValue !== 'string') throw new Error(`FlagInstance's 'usedValue' property must be a string when name is a string`);
        else if (_name instanceof Option && usedValue !== true) throw new Error(`FlagInstance's 'usedValue' property must be \`true\` when name is an Option`);

        this.usedValue = usedValue as TIsOption extends true ? true : string;

        if (document) this.attachDocument(document);
    }

    /** Attaches this flag instance to a document */
    attachDocument(document: Document) {
        console.log('Attaching flag isntance to document', this, document);
        this.documents.add(document);

        let instancesForDoc = FlagInstancesByDocument.get(document);
        if (!instancesForDoc) {
            instancesForDoc = {
                all: new Set(),
                byName: new Map(),
            };
            FlagInstancesByDocument.set(document, instancesForDoc);
        }

        instancesForDoc.all.add(this);

        let instancesByNameSet = instancesForDoc.byName.get(this.name);
        if (!instancesByNameSet) {
            instancesByNameSet = new Set();
            instancesForDoc.byName.set(this.name, instancesByNameSet);
        }

        instancesByNameSet.add(this);
    }

    /** Removes this flag instance from a document */
    removeFromDocument(document: Document) {
        const instancesForDoc = FlagInstancesByDocument.get(document);
        if (!instancesForDoc) {
            this.documents.delete(document);
            return;
        }

        instancesForDoc.all.delete(this);
        instancesForDoc.byName.get(this.name)?.delete(this);

        this.documents.delete(document);
    }

    /** Cleans up the document map */
    delete() {
        this.documents.forEach(document => this.removeFromDocument(document));
    }

    decommission(currentDocument?: Document) {
        currentDocument ? this.removeFromDocument(currentDocument) : this.delete();
    }

    associateWithDocument(document: Document) {
        if (!this.documents.has(document)) this.attachDocument(document);
    }
}
