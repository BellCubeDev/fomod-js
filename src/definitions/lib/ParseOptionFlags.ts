import { FlagDependency, FlagSetter, Install, InstallPattern, Option } from "../module";
import { DefaultFomodDocumentConfig, FomodDocumentConfig } from "./FomodDocumentConfig";

/** Parse all of the option flags in the given options array, setting the flags on the options and resolving any dependencies.
 *
 * This should be called immediately after parsing all of the options in a Fomod document, and should not
 * be called more than once per document.
 */
export function parseOptionFlags(options: Option<boolean>[], document: Document, config: FomodDocumentConfig = {}, dependencies: FlagDependency[] = []) {
    const uniqueFlagSetters = new Map<string, [setter: FlagSetter, option: Option<boolean>, deps: FlagDependency[]]>();
    const setFlagsFound = new Set<string>();

    const loose = (config.parseOptionFlags ?? DefaultFomodDocumentConfig.parseOptionFlags) === 'loose';
    const strictValue = config.optionSelectedValue ?? DefaultFomodDocumentConfig.optionSelectedValue;

    for (const option of options) {
        for (const flag of option.flagsToSet) {
            const uniqueFlag = uniqueFlagSetters.get(flag.name);
            if (uniqueFlag && uniqueFlag[1] !== option) {
                uniqueFlagSetters.delete(flag.name);
                continue;
            }

            if (!setFlagsFound.has(flag.name)) {
                if (loose || flag.value === strictValue) uniqueFlagSetters.set(flag.name, [flag, option, []]);
                setFlagsFound.add(flag.name);
            }
        }
    }

    for (const dep of dependencies) {
        if (dep.flagKey instanceof Option) continue;

        const mapped = uniqueFlagSetters.get(dep.flagKey);
        if (!mapped) continue;
        const [setter, option, deps] = mapped;

        if (setter.value === dep.desiredValue) deps.push(dep);
        else uniqueFlagSetters.delete(dep.flagKey);
    }

    for (const [flagName, [setter, option, deps]] of uniqueFlagSetters) {
        option.flagsToSet.delete(setter);
        option.existingOptionFlagSetterByDocument.set(document, setter);
        setter.flagInstance.optionFlagOptionByDocument.set(document, option);

        for (const dep of deps) {
            dep.flagKey = option;
            dep.desiredValue = true;
        }
    }
}
