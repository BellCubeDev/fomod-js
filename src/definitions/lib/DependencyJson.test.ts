import { DependenciesJsonItem, DependencyType, jsonifyDependencyElements, parseDependencyJsonItems, type DependenciesJsonFlag } from './DependencyJson';
import { assert, beforeEach, describe, expect, test } from 'vitest';
import { parseTag } from '../../tests/testUtils';
import { DependenciesGroup, FlagDependency, GameVersionDependency, ModManagerVersionDependency, Option, ScriptExtenderVersionDependency } from '../module';
import { parseOptionFlags } from './ParseOptionFlags';
import type { FomodDocumentConfig } from './FomodDocumentConfig';
import { DependencyGroupOperator, FileDependencyState, TagName } from '../Enums';
import { FlagInstanceStoresByDocument } from './FlagInstance';

let document: Document;
beforeEach(() => {
    document = parseTag`<p />`.ownerDocument;
});

function createPermutationForFlagDep(deps: DependenciesJsonItem<boolean>[], index: number): DependenciesJsonItem<boolean>[] | null {
    const dep = deps[index];
    if (dep.type !== DependencyType.Flag || typeof dep.flag !== 'string' || !dep.isOptionFlag) return null;
    const newDeps = [...deps];
    newDeps[index] = { ...dep, isOptionFlag: false };
    return newDeps;
}

/** For making sure we parsed stuff correctly, since we can fall back to a string for the flag key if the option doesn't exist,
 * we have to try every permutation of the dependency json with different variations of option flag dependencies set to, well, *not* option flags.
 * That said, non-serializable option flag dependencies shouldn't be able to fall back, hence we don't need to test those permutations.
 *
 * @param deps
 */
function getPermutationsOfDependencyJsonWithOptionFlagMaybeSetToFalse(deps: DependenciesJsonItem<boolean> | DependenciesJsonItem<boolean>[]): DependenciesJsonItem<boolean>[][] {
    const depsArr = Array.isArray(deps) ? deps : [deps];

    const hasOptionFlagDeps = depsArr.some(dep => dep.type === DependencyType.Flag && dep.isOptionFlag);
    if (!hasOptionFlagDeps) return [depsArr];

    const permutations: DependenciesJsonItem<boolean>[][] = [depsArr];

    for (let i = 0; i < depsArr.length; i++) {
        const newPermutations = permutations.map(permutation => createPermutationForFlagDep(permutation, i)).filter((x) => x !== null);
        permutations.push(...newPermutations);
    }

    return permutations;
}

function expect_or(tests: (() => void)[]) {
    if (!tests || !Array.isArray(tests)) return;
    try {
        tests.shift()?.();
        return;
    } catch (e) {
        //console.log('Caught assertion failure; tests left:', tests.length);
        if (tests.length) expect_or(tests);
        else throw e;
    }
  }

function basicTest<TSerializable extends boolean>(deps: DependenciesJsonItem<TSerializable> | DependenciesJsonItem<boolean>[]) {
    const depsArr = Array.isArray(deps) ? deps : [deps];

    const parsedWithDocument = parseDependencyJsonItems(depsArr, document);
    expect(parsedWithDocument).toMatchSnapshot();

    const parsedWithoutDocument = parseDependencyJsonItems(depsArr, null);
    expect(parsedWithoutDocument).toMatchSnapshot();

    const jsonifiedSerializableFromWithout = jsonifyDependencyElements(parsedWithDocument, true, document);
    expect(jsonifiedSerializableFromWithout).toMatchSnapshot();

    const jsonifiedSerializableFromWith = jsonifyDependencyElements(parsedWithDocument, true, document);
    expect(jsonifiedSerializableFromWith).toMatchSnapshot();

    const jsonifiedNonSerializableFromWithout = jsonifyDependencyElements(parsedWithDocument, false);
    expect(jsonifiedNonSerializableFromWithout).toMatchSnapshot();

    const jsonifiedNonSerializableFromWith = jsonifyDependencyElements(parsedWithDocument, false);
    expect(jsonifiedNonSerializableFromWith).toMatchSnapshot();

    const arr = [jsonifiedSerializableFromWithout, jsonifiedSerializableFromWith, jsonifiedNonSerializableFromWithout, jsonifiedNonSerializableFromWith ];
    //expect([jsonifiedSerializableFromWithout, jsonifiedSerializableFromWith, jsonifiedNonSerializableFromWithout, jsonifiedNonSerializableFromWith ]).toContainEqual(getPermutationsOfDependencyJsonWithOptionFlagSetToFalse(depsArr));
    expect_or(
        getPermutationsOfDependencyJsonWithOptionFlagMaybeSetToFalse(depsArr).reverse().map(permutation => (
            () => {
                //console.log('Checking permutation', permutation);
                if (permutation.length === 0) return;

                expect(arr).toContainEqual(permutation);
            }
        ))
    );

    return {
        parsedWithDocument,
        parsedWithoutDocument,
        jsonifiedSerializableFromWithout,
        jsonifiedSerializableFromWith,
        jsonifiedNonSerializableFromWithout,
        jsonifiedNonSerializableFromWith,
    };
}



test('File Dependency', () => {
    const parsedStuff = basicTest({
        type: DependencyType.File,
        path: 'file.txt',
        state: FileDependencyState.Active,
    });
});

test('Flag Dependency (Not Option)', () => {
    const parsedStuff = basicTest({
        type: DependencyType.Flag,
        flag: 'some-flag-name',
        value: 'some-value',
        isOptionFlag: false,
    });
});

test('Flag Dependency (Option, No Existing Option)', () => {
    const parsedStuff = basicTest({
        type: DependencyType.Flag,
        flag: 'some-flag-name',
        value: 'some-value',
        isOptionFlag: true,
    });

    const dep = parsedStuff.parsedWithDocument[0];
    expect(dep).toBeInstanceOf(FlagDependency);
    if (!(dep instanceof FlagDependency)) return;
    assert.typeOf(dep.flagKey, 'string', 'Flag key should fall back to being a string since there is no option to set as the flag key');
});

test('Flag Dependency (Option, Yes Existing Option)', () => {
    const optionElement = parseTag`<plugin name="Test Option">
        <description />
        <conditionFlags>
            <flag name="some-flag-name">OPTION_SELECTED</flag>
        </conditionFlags>
        <files />
        <typeDescriptor>
            <type name="Optional" />
        </typeDescriptor>
    </plugin>`;

    document = optionElement.ownerDocument;

    const parseOptions: FomodDocumentConfig = {
        parseOptionFlags: 'loose',
    };

    const option = Option.parse(optionElement, parseOptions);
    expect(option).not.toBeNull();
    if (!option) return;

    const flagsForDocument1 = FlagInstanceStoresByDocument.get(document);
    expect(flagsForDocument1).not.toBeUndefined();
    if (flagsForDocument1 === undefined) return;
    expect(flagsForDocument1.all.size).toBe(1);
    const flagInstances1 = flagsForDocument1.byName.get('some-flag-name');
    expect(flagInstances1).not.toBeUndefined();
    expect(flagInstances1?.size).toBe(1);

    parseOptionFlags([option], document, parseOptions);

    const flagsForDocument2 = FlagInstanceStoresByDocument.get(document);
    expect(flagsForDocument2).not.toBeUndefined();
    if (flagsForDocument2 === undefined) return;
    expect(flagsForDocument2.all.size).toBe(1);
    const flagInstances2 = flagsForDocument1.byName.get('some-flag-name');
    expect(flagInstances2).not.toBeUndefined();
    expect(flagInstances2?.size).toBe(1);

    const parsedStuff = basicTest({
        type: DependencyType.Flag,
        flag: 'some-flag-name',
        value: 'OPTION_SELECTED',
        isOptionFlag: true,
    });

    const dep = parsedStuff.parsedWithDocument[0];
    assert.strictEqual(document, optionElement.ownerDocument, 'Document should still be the same as the option element\'s document; if not, something went really, really wrong!');
    expect(dep).toBeInstanceOf(FlagDependency);
    if (!(dep instanceof FlagDependency)) return;
    assert.instanceOf(dep.flagKey, Option, 'Flag key should be an Option instance because the option does exist and could be found by the flag key');
});

test('Game Version Dependency', () => {
    const parsedStuff = basicTest({
        type: DependencyType.GameVersion,
        version: '1.2.3',
    });

    const dep = parsedStuff.parsedWithDocument[0];
    expect(dep).toBeInstanceOf(GameVersionDependency);
});

test('Group of Dependencies (No Children)', () => {
    const parsedStuff = basicTest({
        type: DependencyType.Group,
        operator: DependencyGroupOperator.And,
        tagName: TagName.Dependencies,
        dependencies: []
    });

    const group = parsedStuff.parsedWithDocument[0];
    expect(group).toBeInstanceOf(DependenciesGroup);
    if (!(group instanceof DependenciesGroup)) return;
    assert.isEmpty(group.dependencies, 'This group should have no dependencies');
});

test('Group of Dependencies (With Children)', () => {
    const parsedStuff = basicTest({
        type: DependencyType.Group,
        operator: DependencyGroupOperator.And,
        tagName: TagName.Dependencies,
        dependencies: [{
            type: DependencyType.GameVersion,
            version: '1.2.3',
        }]
    });

    const group = parsedStuff.parsedWithDocument[0];
    expect(group).toBeInstanceOf(DependenciesGroup);
    if (!(group instanceof DependenciesGroup)) return;
    assert.isNotEmpty(group.dependencies, 'This group should have dependencies');
    assert.equal(group.dependencies.size, 1, 'This group should have exactly one dependency');
});

test('Mod Manager Version Dependency', () => {
    const parsedStuff = basicTest({
        type: DependencyType.ModManagerVersion,
        version: '1.2.3',
    });

    const dep = parsedStuff.parsedWithDocument[0];
    expect(dep).toBeInstanceOf(ModManagerVersionDependency);
});

test('Script Extender Version Dependency', () => {
    const parsedStuff = basicTest({
        type: DependencyType.ScriptExtenderVersion,
        version: '1.2.3',
    });

    const dep = parsedStuff.parsedWithDocument[0];
    expect(dep).toBeInstanceOf(ScriptExtenderVersionDependency);
});
