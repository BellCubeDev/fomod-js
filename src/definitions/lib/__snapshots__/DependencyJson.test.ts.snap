// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`File Dependency 1`] = `
[
  FileDependency {
    "decommission": undefined,
    "desiredState": "Active",
    "documentMap": WeakMap {},
    "filePath": "file.txt",
    "tagName": "fileDependency",
  },
]
`;

exports[`File Dependency 2`] = `
[
  FileDependency {
    "decommission": undefined,
    "desiredState": "Active",
    "documentMap": WeakMap {},
    "filePath": "file.txt",
    "tagName": "fileDependency",
  },
]
`;

exports[`File Dependency 3`] = `
[
  {
    "path": "file.txt",
    "state": "Active",
    "type": "file",
  },
]
`;

exports[`File Dependency 4`] = `
[
  {
    "path": "file.txt",
    "state": "Active",
    "type": "file",
  },
]
`;

exports[`File Dependency 5`] = `
[
  {
    "path": "file.txt",
    "state": "Active",
    "type": "file",
  },
]
`;

exports[`File Dependency 6`] = `
[
  {
    "path": "file.txt",
    "state": "Active",
    "type": "file",
  },
]
`;

exports[`Flag Dependency (Not Option) 1`] = `
[
  FlagDependency {
    "documentMap": WeakMap {},
    "flagInstance": FlagInstance {
      "_name": "some-flag-name",
      "documents": Set {},
      "optionFlagOptionByDocument": WeakMap {},
      "usedValue": "some-value",
      "write": false,
    },
    "tagName": "flagDependency",
  },
]
`;

exports[`Flag Dependency (Not Option) 2`] = `
[
  FlagDependency {
    "documentMap": WeakMap {},
    "flagInstance": FlagInstance {
      "_name": "some-flag-name",
      "documents": Set {},
      "optionFlagOptionByDocument": WeakMap {},
      "usedValue": "some-value",
      "write": false,
    },
    "tagName": "flagDependency",
  },
]
`;

exports[`Flag Dependency (Not Option) 3`] = `
[
  {
    "flag": "some-flag-name",
    "isOptionFlag": false,
    "type": "flag",
    "value": "some-value",
  },
]
`;

exports[`Flag Dependency (Not Option) 4`] = `
[
  {
    "flag": "some-flag-name",
    "isOptionFlag": false,
    "type": "flag",
    "value": "some-value",
  },
]
`;

exports[`Flag Dependency (Not Option) 5`] = `
[
  {
    "flag": "some-flag-name",
    "isOptionFlag": false,
    "type": "flag",
    "value": "some-value",
  },
]
`;

exports[`Flag Dependency (Not Option) 6`] = `
[
  {
    "flag": "some-flag-name",
    "isOptionFlag": false,
    "type": "flag",
    "value": "some-value",
  },
]
`;

exports[`Flag Dependency (Option, No Existing Option) 1`] = `
[
  FlagDependency {
    "documentMap": WeakMap {},
    "flagInstance": FlagInstance {
      "_name": "some-flag-name",
      "documents": Set {},
      "optionFlagOptionByDocument": WeakMap {},
      "usedValue": "some-value",
      "write": false,
    },
    "tagName": "flagDependency",
  },
]
`;

exports[`Flag Dependency (Option, No Existing Option) 2`] = `
[
  FlagDependency {
    "documentMap": WeakMap {},
    "flagInstance": FlagInstance {
      "_name": "some-flag-name",
      "documents": Set {},
      "optionFlagOptionByDocument": WeakMap {},
      "usedValue": "some-value",
      "write": false,
    },
    "tagName": "flagDependency",
  },
]
`;

exports[`Flag Dependency (Option, No Existing Option) 3`] = `
[
  {
    "flag": "some-flag-name",
    "isOptionFlag": false,
    "type": "flag",
    "value": "some-value",
  },
]
`;

exports[`Flag Dependency (Option, No Existing Option) 4`] = `
[
  {
    "flag": "some-flag-name",
    "isOptionFlag": false,
    "type": "flag",
    "value": "some-value",
  },
]
`;

exports[`Flag Dependency (Option, No Existing Option) 5`] = `
[
  {
    "flag": "some-flag-name",
    "isOptionFlag": false,
    "type": "flag",
    "value": "some-value",
  },
]
`;

exports[`Flag Dependency (Option, No Existing Option) 6`] = `
[
  {
    "flag": "some-flag-name",
    "isOptionFlag": false,
    "type": "flag",
    "value": "some-value",
  },
]
`;

exports[`Flag Dependency (Option, Yes Existing Option) 1`] = `
[
  FlagDependency {
    "documentMap": WeakMap {},
    "flagInstance": FlagInstance {
      "_name": Option {
        "description": "",
        "documentMap": WeakMap {},
        "existingOptionFlagSetterByDocument": WeakMap {},
        "flagsToSet": Set {},
        "image": null,
        "installsToSet": InstallPattern {
          "dependencies": DependenciesGroup {
            "dependencies": Set {
              FlagDependency {
                "documentMap": WeakMap {},
                "flagInstance": FlagInstance {
                  "_name": [Circular],
                  "documents": Set {},
                  "optionFlagOptionByDocument": WeakMap {},
                  "usedValue": true,
                  "write": false,
                },
                "tagName": "flagDependency",
              },
            },
            "documentMap": WeakMap {},
            "operator": "And",
            "tagName": "dependencies",
          },
          "documentMap": WeakMap {},
          "filesWrapper": InstallPatternFilesWrapper {
            "documentMap": WeakMap {},
            "installs": Set {},
            "tagName": "files",
          },
          "tagName": "pattern",
        },
        "name": "Test Option",
        "tagName": "plugin",
        "typeDescriptor": TypeDescriptor {
          "defaultTypeNameDescriptor": TypeNameDescriptor {
            "_tagName": "type",
            "documentMap": WeakMap {},
            "tagNameIsReadonly": undefined,
            "targetType": "Optional",
          },
          "documentMap": WeakMap {},
          "patterns": [],
          "tagName": "typeDescriptor",
        },
      },
      "documents": Set {},
      "optionFlagOptionByDocument": WeakMap {},
      "usedValue": true,
      "write": false,
    },
    "tagName": "flagDependency",
  },
]
`;

exports[`Flag Dependency (Option, Yes Existing Option) 2`] = `
[
  FlagDependency {
    "documentMap": WeakMap {},
    "flagInstance": FlagInstance {
      "_name": "some-flag-name",
      "documents": Set {},
      "optionFlagOptionByDocument": WeakMap {},
      "usedValue": "OPTION_SELECTED",
      "write": false,
    },
    "tagName": "flagDependency",
  },
]
`;

exports[`Flag Dependency (Option, Yes Existing Option) 3`] = `
[
  {
    "flag": "some-flag-name",
    "isOptionFlag": true,
    "type": "flag",
    "value": "OPTION_SELECTED",
  },
]
`;

exports[`Flag Dependency (Option, Yes Existing Option) 4`] = `
[
  {
    "flag": "some-flag-name",
    "isOptionFlag": true,
    "type": "flag",
    "value": "OPTION_SELECTED",
  },
]
`;

exports[`Flag Dependency (Option, Yes Existing Option) 5`] = `
[
  {
    "flag": Option {
      "description": "",
      "documentMap": WeakMap {},
      "existingOptionFlagSetterByDocument": WeakMap {},
      "flagsToSet": Set {},
      "image": null,
      "installsToSet": InstallPattern {
        "dependencies": DependenciesGroup {
          "dependencies": Set {
            FlagDependency {
              "documentMap": WeakMap {},
              "flagInstance": FlagInstance {
                "_name": [Circular],
                "documents": Set {},
                "optionFlagOptionByDocument": WeakMap {},
                "usedValue": true,
                "write": false,
              },
              "tagName": "flagDependency",
            },
          },
          "documentMap": WeakMap {},
          "operator": "And",
          "tagName": "dependencies",
        },
        "documentMap": WeakMap {},
        "filesWrapper": InstallPatternFilesWrapper {
          "documentMap": WeakMap {},
          "installs": Set {},
          "tagName": "files",
        },
        "tagName": "pattern",
      },
      "name": "Test Option",
      "tagName": "plugin",
      "typeDescriptor": TypeDescriptor {
        "defaultTypeNameDescriptor": TypeNameDescriptor {
          "_tagName": "type",
          "documentMap": WeakMap {},
          "tagNameIsReadonly": undefined,
          "targetType": "Optional",
        },
        "documentMap": WeakMap {},
        "patterns": [],
        "tagName": "typeDescriptor",
      },
    },
    "isOptionFlag": true,
    "type": "flag",
    "value": true,
  },
]
`;

exports[`Flag Dependency (Option, Yes Existing Option) 6`] = `
[
  {
    "flag": Option {
      "description": "",
      "documentMap": WeakMap {},
      "existingOptionFlagSetterByDocument": WeakMap {},
      "flagsToSet": Set {},
      "image": null,
      "installsToSet": InstallPattern {
        "dependencies": DependenciesGroup {
          "dependencies": Set {
            FlagDependency {
              "documentMap": WeakMap {},
              "flagInstance": FlagInstance {
                "_name": [Circular],
                "documents": Set {},
                "optionFlagOptionByDocument": WeakMap {},
                "usedValue": true,
                "write": false,
              },
              "tagName": "flagDependency",
            },
          },
          "documentMap": WeakMap {},
          "operator": "And",
          "tagName": "dependencies",
        },
        "documentMap": WeakMap {},
        "filesWrapper": InstallPatternFilesWrapper {
          "documentMap": WeakMap {},
          "installs": Set {},
          "tagName": "files",
        },
        "tagName": "pattern",
      },
      "name": "Test Option",
      "tagName": "plugin",
      "typeDescriptor": TypeDescriptor {
        "defaultTypeNameDescriptor": TypeNameDescriptor {
          "_tagName": "type",
          "documentMap": WeakMap {},
          "tagNameIsReadonly": undefined,
          "targetType": "Optional",
        },
        "documentMap": WeakMap {},
        "patterns": [],
        "tagName": "typeDescriptor",
      },
    },
    "isOptionFlag": true,
    "type": "flag",
    "value": true,
  },
]
`;

exports[`Game Version Dependency 1`] = `
[
  GameVersionDependency {
    "decommission": undefined,
    "desiredVersion": "1.2.3",
    "documentMap": WeakMap {},
    "tagName": "gameDependency",
  },
]
`;

exports[`Game Version Dependency 2`] = `
[
  GameVersionDependency {
    "decommission": undefined,
    "desiredVersion": "1.2.3",
    "documentMap": WeakMap {},
    "tagName": "gameDependency",
  },
]
`;

exports[`Game Version Dependency 3`] = `
[
  {
    "type": "game",
    "version": "1.2.3",
  },
]
`;

exports[`Game Version Dependency 4`] = `
[
  {
    "type": "game",
    "version": "1.2.3",
  },
]
`;

exports[`Game Version Dependency 5`] = `
[
  {
    "type": "game",
    "version": "1.2.3",
  },
]
`;

exports[`Game Version Dependency 6`] = `
[
  {
    "type": "game",
    "version": "1.2.3",
  },
]
`;

exports[`Group of Dependencies (No Children) 1`] = `
[
  DependenciesGroup {
    "dependencies": Set {},
    "documentMap": WeakMap {},
    "operator": "And",
    "tagName": "dependencies",
  },
]
`;

exports[`Group of Dependencies (No Children) 2`] = `
[
  DependenciesGroup {
    "dependencies": Set {},
    "documentMap": WeakMap {},
    "operator": "And",
    "tagName": "dependencies",
  },
]
`;

exports[`Group of Dependencies (No Children) 3`] = `
[
  {
    "dependencies": [],
    "operator": "And",
    "tagName": "dependencies",
    "type": "group",
  },
]
`;

exports[`Group of Dependencies (No Children) 4`] = `
[
  {
    "dependencies": [],
    "operator": "And",
    "tagName": "dependencies",
    "type": "group",
  },
]
`;

exports[`Group of Dependencies (No Children) 5`] = `
[
  {
    "dependencies": [],
    "operator": "And",
    "tagName": "dependencies",
    "type": "group",
  },
]
`;

exports[`Group of Dependencies (No Children) 6`] = `
[
  {
    "dependencies": [],
    "operator": "And",
    "tagName": "dependencies",
    "type": "group",
  },
]
`;

exports[`Group of Dependencies (With Children) 1`] = `
[
  DependenciesGroup {
    "dependencies": Set {
      GameVersionDependency {
        "decommission": undefined,
        "desiredVersion": "1.2.3",
        "documentMap": WeakMap {},
        "tagName": "gameDependency",
      },
    },
    "documentMap": WeakMap {},
    "operator": "And",
    "tagName": "dependencies",
  },
]
`;

exports[`Group of Dependencies (With Children) 2`] = `
[
  DependenciesGroup {
    "dependencies": Set {
      GameVersionDependency {
        "decommission": undefined,
        "desiredVersion": "1.2.3",
        "documentMap": WeakMap {},
        "tagName": "gameDependency",
      },
    },
    "documentMap": WeakMap {},
    "operator": "And",
    "tagName": "dependencies",
  },
]
`;

exports[`Group of Dependencies (With Children) 3`] = `
[
  {
    "dependencies": [
      {
        "type": "game",
        "version": "1.2.3",
      },
    ],
    "operator": "And",
    "tagName": "dependencies",
    "type": "group",
  },
]
`;

exports[`Group of Dependencies (With Children) 4`] = `
[
  {
    "dependencies": [
      {
        "type": "game",
        "version": "1.2.3",
      },
    ],
    "operator": "And",
    "tagName": "dependencies",
    "type": "group",
  },
]
`;

exports[`Group of Dependencies (With Children) 5`] = `
[
  {
    "dependencies": [
      {
        "type": "game",
        "version": "1.2.3",
      },
    ],
    "operator": "And",
    "tagName": "dependencies",
    "type": "group",
  },
]
`;

exports[`Group of Dependencies (With Children) 6`] = `
[
  {
    "dependencies": [
      {
        "type": "game",
        "version": "1.2.3",
      },
    ],
    "operator": "And",
    "tagName": "dependencies",
    "type": "group",
  },
]
`;

exports[`Mod Manager Version Dependency 1`] = `
[
  ModManagerVersionDependency {
    "decommission": undefined,
    "desiredVersion": "1.2.3",
    "documentMap": WeakMap {},
    "tagName": "fommDependency",
  },
]
`;

exports[`Mod Manager Version Dependency 2`] = `
[
  ModManagerVersionDependency {
    "decommission": undefined,
    "desiredVersion": "1.2.3",
    "documentMap": WeakMap {},
    "tagName": "fommDependency",
  },
]
`;

exports[`Mod Manager Version Dependency 3`] = `
[
  {
    "type": "fomm",
    "version": "1.2.3",
  },
]
`;

exports[`Mod Manager Version Dependency 4`] = `
[
  {
    "type": "fomm",
    "version": "1.2.3",
  },
]
`;

exports[`Mod Manager Version Dependency 5`] = `
[
  {
    "type": "fomm",
    "version": "1.2.3",
  },
]
`;

exports[`Mod Manager Version Dependency 6`] = `
[
  {
    "type": "fomm",
    "version": "1.2.3",
  },
]
`;

exports[`Script Extender Version Dependency 1`] = `
[
  ScriptExtenderVersionDependency {
    "decommission": undefined,
    "desiredVersion": "1.2.3",
    "documentMap": WeakMap {},
    "tagName": "foseDependency",
  },
]
`;

exports[`Script Extender Version Dependency 2`] = `
[
  ScriptExtenderVersionDependency {
    "decommission": undefined,
    "desiredVersion": "1.2.3",
    "documentMap": WeakMap {},
    "tagName": "foseDependency",
  },
]
`;

exports[`Script Extender Version Dependency 3`] = `
[
  {
    "type": "fose",
    "version": "1.2.3",
  },
]
`;

exports[`Script Extender Version Dependency 4`] = `
[
  {
    "type": "fose",
    "version": "1.2.3",
  },
]
`;

exports[`Script Extender Version Dependency 5`] = `
[
  {
    "type": "fose",
    "version": "1.2.3",
  },
]
`;

exports[`Script Extender Version Dependency 6`] = `
[
  {
    "type": "fose",
    "version": "1.2.3",
  },
]
`;
